File: go/ordered_test.go
/* Copyright (c) 2026 Richard Rodger, MIT License */

// Key-order parity with the TS port's `map: { ordered: true }` mode.
//
// The Go port's *OrderedMap preserves insertion order for every key by
// construction. The TS port returns plain JS objects, which reorder
// integer-like keys ascending at creation; its opt-in ordered mode
// records insertion order in a side channel read by `keyOrder`. The
// TABLE below is duplicated verbatim in ts/test/ordered.test.js — the
// same inputs must report the same key order in both ports. (A shared
// TSV cannot pin this: its comparison is a JSON round-trip, and
// JSON.stringify re-loses integer-key order — the thing under test.)

package tabnasjsonic

import (
	"reflect"
	"testing"
)

// KEEP IN SYNC with ts/test/ordered.test.js TABLE.
var orderedCases = []struct {
	src   string
	order []string
}{
	{"{2:9, 1:8}", []string{"2", "1"}},
	{"{10:a, 2:b, x:c}", []string{"10", "2", "x"}},
	{"{a:1, 2:b, a:3}", []string{"a", "2"}}, // repeated key keeps first position
	{"{zz:1, 0:2, aa:3}", []string{"zz", "0", "aa"}},
}

func TestOrderedKeyParity(t *testing.T) {
	for _, c := range orderedCases {
		j := Make()
		v, err := j.Parse(c.src)
		if err != nil {
			t.Fatalf("Parse(%q): %v", c.src, err)
		}
		om, ok := v.(*OrderedMap)
		if !ok {
			t.Fatalf("Parse(%q): expected *OrderedMap, got %T", c.src, v)
		}
		if !reflect.DeepEqual(om.Keys, c.order) {
			t.Errorf("Parse(%q) key order = %v, want %v (cross-port table)",
				c.src, om.Keys, c.order)
		}
	}
}

func TestOrderedKeyNestedMerge(t *testing.T) {
	j := Make()
	v, err := j.Parse("a:b:1,a:c:2")
	if err != nil {
		t.Fatal(err)
	}
	a, _ := v.(*OrderedMap).Get("a")
	if got := a.(*OrderedMap).Keys; !reflect.DeepEqual(got, []string{"b", "c"}) {
		t.Errorf("merged nested key order = %v, want [b c]", got)
	}
}

File: ts/test/ordered.test.js
/* Copyright (c) 2026 Richard Rodger, MIT License */
'use strict'

// map.ordered — the TS mirror of the Go port's OrderedMap.
//
// Plain JS objects enumerate integer-like keys in ascending numeric order
// no matter the order the user wrote them — a language semantic, not a
// bug — so `{2:9,1:8}` irrecoverably loses its 2-before-1 key order at
// object creation, while the Go port's *jsonic.OrderedMap preserves it
// for every key. `map: { ordered: true }` records insertion order in a
// non-enumerable side channel read by `keyOrder`, restoring cross-port
// parity without changing the plain-object result shape.
//
// The TABLE below is duplicated verbatim in go/ordered_test.go, asserting
// OrderedMap.Keys — the same inputs must report the same key order in
// both ports. (A shared TSV cannot pin this: the comparison there is a
// JSON round-trip, and JSON.stringify re-loses integer-key order — the
// very thing under test.)

const { describe, it } = require('node:test')
const assert = require('node:assert')

const { Jsonic, keyOrder } = require('..')

// KEEP IN SYNC with go/ordered_test.go orderedCases.
const TABLE = [
  { src: '{2:9, 1:8}', order: ['2', '1'] },
  { src: '{10:a, 2:b, x:c}', order: ['10', '2', 'x'] },
  { src: '{a:1, 2:b, a:3}', order: ['a', '2'] }, // repeated key keeps first position
  { src: '{zz:1, 0:2, aa:3}', order: ['zz', '0', 'aa'] },
]

describe('ordered', () => {
  const jo = Jsonic.make({ map: { ordered: true } })

  it('keyOrder reports source insertion order (cross-port table)', () => {
    for (const { src, order } of TABLE) {
      assert.deepStrictEqual(keyOrder(jo.parse(src)), order, src)
    }
  })

  it('nested and deep-merged maps record order too', () => {
    assert.deepStrictEqual(keyOrder(jo.parse('a:b:1,a:c:2').a), ['b', 'c'])
    assert.deepStrictEqual(keyOrder(jo.parse('{9:{z:1}, 2:x}')['9']), ['z'])
  })

  it('the result is still a plain object — no consumer breakage', () => {
    const r = jo.parse('{2:9, 1:8}')
    assert.strictEqual(Object.getPrototypeOf(r), null)
    assert.deepStrictEqual(JSON.parse(JSON.stringify(r)), { 1: 8, 2: 9 })
    // The record is invisible to enumeration.
    assert.deepStrictEqual(Object.keys(r), ['1', '2'])
  })

  it('plain-object mode DOCUMENTED LIMITATION: integer-like key order is lost', () => {
    // Without map.ordered there is no record, and keyOrder falls back to
    // Object.keys — ascending integer order, regardless of the source.
    // This is JS enumeration semantics; the plain representation cannot
    // carry the information. Pinned so the limitation stays documented.
    const r = Jsonic.make().parse('{2:9, 1:8}')
    assert.deepStrictEqual(keyOrder(r), ['1', '2'])
  })
})
